Alternative library restructure

howeik opened the pull request:
My alternative to https://github.com/octokit/octokit.py/pull/18 to address https://github.com/octokit/octokit.py/issues/17

Highlights
- Resource remains the hypermedia agent.
- Client is a hypermedia agent (subclass of Resource) that additionally contains octokit specific fields, functionality, and mixins. This allows Client to act as a specialized hypermedia agent.
- Because of the above, octokit specific code can now be decoupled from Resource and coupled with Client instead.
- Moved Client functionality to BaseClient, making Client just a mixin of BaseClient, RateLimit, and future mixin classes. This allows each mixin to have its constructor called so we don't have to initialize mixin fields lazily or in Client's constructor.
- We're able to paginate all requests automatically (hypermedia, or through a convenience class)
- We're able to handle octokit specific exceptions independently of Resource

/cc @octokit/ucsd-octokit-py

howeik committed:
Added RateLimit mixin + response hook to update last_response

howeik committed:
Deleted custom Session class. Moved pagination logic to Client and updated pagination tests.

eduardoramirez commented on octokit/client.py:
```diff
@@ -28,8 +29,43 @@ class Client(Resource):
     self.schema = {}
     self.name = 'Client'
+    self._rate_limit = _RateLimit()
+    self.auto_paginate = True
```
Not sure if this should be initiated to True by default?

eduardoramirez commented on octokit/client.py:
```diff
@@ -28,8 +29,43 @@ class Client(Resource):
+
+  def response_callback(self, r, *args, **kwargs):
+    self.last_response = r
+    # TODO (howei): perhaps we could auto-paginate requests here
```
This is interesting.. we wouldn't need to directly call paginate ever.

howeik replied:
Yep. We could also address one of your TODOs in agent by handling HTTP response status codes here instead of in Resource (since it's octokit specific).

eduardoramirez commented:
Looks good! If we are focusing on hypermedia only, this should work

eduardoramirez commented on octokit/client.py:
```diff
@@ -9,11 +9,12 @@
+import requests
 
-class Client(Resource):
+class Client(Resource, RateLimit):
```
I think this should be reverse `Client(RateLimit, Resource)`. Otherwise, Client is a RateLimit instead of a resource

howeik commented:
Thanks! :zzz: 
It should allow for our convenience mixins too.

howeik committed:
Handle HTTP response codes in Client instead of Resource.

howeik committed:
Set auto_paginate to False by default; fixed pagination tests.

howeik committed:
Moved Client functionality to BaseClient to allow for mixins to call their own constructor.

howeik committed:
Newline at end of client.py

howeik committed:
Moved Client class comments to BaseClient.

howeik committed:
Use `head` http method when requesting rate limit.

howeik committed:
Refactored pagination + Resource/rels relationship

howeik commented:
@eduardoramirez I noticed `Resource.fetch_resource` was assigning the fetched resource's rels to the resource doing the fetching. I think instead we should be assigning the fetched resource's rels to the fetched resource. I've refactored Resource + the pagination logic to to reflect this but think you should take a look: https://github.com/octokit/octokit.py/commit/01557af867485633e38e8750e380b99ec14e7288

eduardoramirez commented:
@howeik Yep, I had fixed that in the other PR

eduardoramirez commented on octokit/resources.py:
```diff
@@ -19,23 +17,26 @@ class Resource(object):
       return self.schema[name]
     else:
-      raise handle_status(404)
+      raise AttributeError
```
I feel like this should still be 404, we could override it in client

eduardoramirez commented on octokit/resources.py:
```diff
@@ -226,7 +196,5 @@ def fetch_resource(self, method, *args, **kwargs):
-    self.session.last_response = response
 
-    return Resource(self.session, schema=schema, name=humanize(self.name))
+    return Resource(self.session, response=response, schema=schema, name=humanize(self.name))
```
I think since we are passing the response in as parameter, we should just move the `self.parse_schema(response)` to the constructor

howeik replied:
Yep me too. I tried to do this though and I ended up in an endless recursion in `__getattr__`. I saw you passed response in your branch though so maybe you know how we could deal with that!

eduardoramirez replied:
This will also clean up the code a little bit and Yea, we can deal with afterwards its not a big deal

eduardoramirez commented:
Just a few nits, but looks good overall! :100:

howeik committed:
Moved schema parsing to constructor of Resource.

howeik committed:
Made default behavior to bad attributes in Client raise a 404 exception.

howeik committed:
Refactored parse_schema funcs in Resource. Set url from response in Resource.

eduardoramirez commented on octokit/ratelimit.py:
```diff
@@ -0,0 +1,40 @@
+    # TODO (eduardo) : Find a better way to join url with api_endpoint so we
+    # don't need to do it for every REST call
+    endpoint = urljoin(self.agent.endpoint, 'rate_limit')
+    self.last_response = self.head(endpoint).response
```
This is leaky from my previous PR, but instead we can do `self.rate_limit` to follow hypermedia .. ahh wait no, that's a property

howeik committed:
Moved pagination logic to Pagination mixin for Client.

howeik committed:
Deleted accidental file add..

howeik committed:
Moved pagination and ratelimit tests to their own test files

eduardoramirez commented on octokit/pagination.py:
```diff
@@ -0,0 +1,38 @@
+    # TODO (howei): possibly extract auto_paginate from kwargs
+    # so users can do client = Client(auto_paginate=True)
+    self.auto_paginate = False
+    self.last_response = None
```
I think last reponse makes more sense in Client.. pagination doesn't use it, and other mixins do (like rate_limit)

howeik replied:
oops no idea why I did that. I was going to put that in ratelimit.py instead of client.py, since client doesn't directly use it and ratelimit seems to be the only thing that does. thoughts?

eduardoramirez replied:
Actually I see that only the rate limit response callback is setting the last response, but we want to keep track of the last response at all times not just when rate limit fetches the rate limit end point?

howeik replied:
Yep because of how we set up mixins, each of the mixin's response_callbacks will be called in the order our mixins are declared:

`class Client(RateLimit, Pagination, ..)` - RateLimit's response callback will be called, then Pagination's then so forth and this will happen with every request because of the hook in session.

Since we can do that I thought it'd make more sense for RateLimit to initialize and and update last_response (as opposed to Client doing this).

eduardoramirez replied:
Hmm okay, I guess atm that makes sense (y)

eduardoramirez commented on octokit/pagination.py:
```diff
@@ -0,0 +1,38 @@
+        resource = resource.rels['next'].get()
+        data.extend(list(resource.schema))
+
+    return Resource(session, schema=data, url=self.url, name=self.name)
```
newline

howeik replied:
done

eduardoramirez commented on octokit/pagination.py:
```diff
@@ -0,0 +1,38 @@
+
+  def response_callback(self, r, **kwargs):
+    # TODO (howei): possibly implement auto pagination logic here
+    return super(Pagination, self).response_callback(r, **kwargs)
```
Why do we have the response callback also here?

howeik replied:
To mark out a place where we could possibly implement api-wide pagination.

eduardoramirez commented:
Let's try to merge this soon, it looks good :+1

howeik committed:
Newline at end of pagination.py

howeik commented:
Yep @tchangg this all look good to you?

howeik committed:
Refactored RateLimit code, added __repr__ for _RateLimit

>>> Client().rate_limit
<class 'octokit.ratelimit._RateLimit'>({'limit': 60, 'resets_at': 1447366754, 'remaining': 38, 'resets_in': 2837})>

howeik committed:
Moved last_response from pagination to ratelimit

howeik committed:
Fixed __slots__ and __repr__ in RateLimit.

>>> Client().rate_limit
<class 'octokit.ratelimit._RateLimit'>(limit=60, remaining=31, resets_at=1447366754, resets_in=1907)>

howeik committed:
Fixed typo in _RateLimit.__repr__

tchangg commented:
LGTM
